### DISCUSSION TITLE
Configuration for new MLCAPE in the new beta

### DISCUSSION THREAD

**Discussion by PerryShafran-NOAA** [dtcenter/METplus#1472] [Configuration] [answered]
Hi, all,

I have been asked to test the new capability in https://github.com/dtcenter/MET/issues/1824 to test for Mixed-Layer CAPE (MLCAPE).  I am unsure of the settings to use in the configuration for PB2NC and PointStat to be able to test, since this is a new capability I haven't used before.  Can someone assist with the settings here?

Thanks!

Perry

**Comment by TaraJensen** [comment]
@hsoh-u - This is time sensitive.  Can you help Perry out?

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA, I saw your email to @jprestop and I, but figure I'd followup here.

Turning on the output for MLCAPE in the PB2NC config file should be pretty straight-forward. Here's an [example](https://github.com/dtcenter/MET/blob/591cbd26ff2f4e3df5ea0354df5ce626b050cad9/test/config/PB2NCConfig_pbl#L95) where @hsoh-u added it to a config file run in MET's unit tests:
```
obs_bufr_var = ["D_CAPE", "D_PBL", "D_MLCAPE"];
```

The "D_MLCAPE" tells PB2NC to derive it. Also note, that this [line](https://github.com/dtcenter/MET/blob/591cbd26ff2f4e3df5ea0354df5ce626b050cad9/met/data/config/PB2NCConfig_default#L127) has been added to the default PB2NC config file:
```
obs_prepbufr_map = [
...
   { key = "D_MLCAPE"; val = "MLCAPE"; }
];
```

So you request it as "D_MLCAPE" in "obs_bufr_var" and you should see observations of "MLCAPE" show up in the output. Increasing the verbosity level should print more details about the derivation. I'm not sure what verbosity level to recommend though.

**Comment by PerryShafran-NOAA** [comment]
Thanks, John!  Can you also assist with the settings in PointStat?  Did you use an NCEP model file such as NAM or RAP?  

Right now, the setting for surface-based CAPE is:

FCST_VAR11_NAME = CAPE
FCST_VAR11_LEVELS = L0
FCST_VAR11_OPTIONS = cnt_thresh = [ >0 ];
FCST_VAR11_THRESH = >500, >1000, >1500, >2000, >3000, >4000
OBS_VAR11_NAME = CAPE
OBS_VAR11_LEVELS = L0-100000
OBS_VAR11_OPTIONS = cnt_thresh = [ >0 ]; cnt_logic = UNION;
OBS_VAR11_THRESH = >500, >1000, >1500, >2000, >3000, >4000

What changes for MLCAPE?

Thanks!

Perry

**Reply by PerryShafran-NOAA** [reply]
Also - I am currently running PB2NC only to test the new capability.  I have verbosity at 7 and that gives some information.  I am not finding any derivations that are for MLCAPE.   How can I tell in the log file which derivations are for CAPE and which are for MLCAPE?  I see CAPE being derived, but not MLCAPE.

Perry

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA I have not actually run Point-Stat to verify MLCAPE. If your model output file contains Mixed Layer CAPE data, and its named MLCAPE in the GRIB1/2 output file, it should be as easy as adding a new entry to your METplus config file. I assumed you'd want to use the same setup as you're using for CAPE?

```
OBS_VAR12_NAME = MLCAPE
OBS_VAR12_LEVELS = L0-100000
OBS_VAR12_OPTIONS = cnt_thresh = [ >0 ]; cnt_logic = UNION;
OBS_VAR12_THRESH = >500, >1000, >1500, >2000, >3000, >4000
```

**Reply by PerryShafran-NOAA** [reply]
Interestingly, there is no grib value output of MLCAPE.  I have the following types of CAPE (this is from RAP):

230:11382185:vt=2022030301:surface:2 hour fcst:CAPE Convective Available Potential Energy [J/kg] - this is surface-based CAPE
289:15257835:vt=2022030301:180-0 mb above ground:2 hour fcst:CAPE Convective Available Potential Energy [J/kg
292:15507077:vt=2022030301:90-0 mb above ground:2 hour fcst:CAPE Convective Available Potential Energy [J/kg]
294:15541945:vt=2022030301:255-0 mb above ground:2 hour fcst:CAPE Convective Available Potential Energy [J/kg]
299:15810877:vt=2022030301:0-3000 m above ground:2 hour fcst:CAPE Convective Available Potential Energy [J/kg]

As you see, I have many different layers to choose from.  Which is the one I should be using, and how would I post that in PointStat?

Thanks!

Perry

**Reply by PerryShafran-NOAA** [reply]
But the more important thing is what I should be looking for in the PB2NC output to know that this worked.  Because I'm not finding anything specific to MLCAPE derivation in the PB2NC output.

**Reply by hsoh-u** [reply]
@[hertneky](https://github.com/hertneky) said the 0-90 mb layer for MLCAPE (UPP)

**Reply by PerryShafran-NOAA** [reply]
@hsoh-u @JohnHalleyGotway Yep, got that info from colleagues here.  

Now how would I read that particular record in PointStat?  I'm not sure how to write the FCST_VAR_LEVELS in PointStat to be sure I am reading the correct record.

**Comment by hsoh-u** [comment, marked as answer]
[log messages]
pb2nc does not log messages for mlcape. The number of CAPE and MLCAPE data is logged with the debug level 3.

Case 1: D_CAPE only
Case 2: D_MLCAPE only
Case 3: Both D_CAPE and D_MLCAPE
```
log_cape_only:DEBUG 3: Derived CAPE = 184      Zero = 62

log_mlcape_only:DEBUG 3: Derived CAPE = 184       Zero = 84

log_both_capes:DEBUG 3: Derived CAPE = 368   Zero = 146
```
[R script to see the MLCAPE]
```
Rscript  met/share/met/Rscripts/pntnc2ascii.R prepda.2018091100.add_cape.nc | grep MLCAPE
ADPUPA 78988 20180911_000000 12.200  -68.970    54 MLCAPE NA     0.00000 2   1398.900000
ADPUPA 80001 20180911_000000 12.580  -81.700     2 MLCAPE NA     0.00000 2   1510.900000
ADPUPA 78954 20180911_000000 13.070  -59.500    47 MLCAPE NA     0.00000 2   1343.400000
ADPUPA 78384 20180911_000000 19.300  -81.370     3 MLCAPE NA     0.00000 2   2281.200000
...
```

**Reply by PerryShafran-NOAA** [reply]
Thank you!  

Well I see that for the CAPE case only I have:

DEBUG 3: Derived CAPE = 95      Zero = 69

And for the CAPE and MLCAPE case, I have:

DEBUG 3: Derived CAPE = 190     Zero = 147

So it looks like we're good - there are obviously more CAPEs calculated with the MLCAPE included.  

I'm trying to find Rscript on WCOSS and can't seem to find it.  Once I see that, then I can validate that the new PB2NC works for calculating MLCAPE.

**Reply by hsoh-u** [reply]
The R script is available from github without building MET:
```
wget https://raw.githubusercontent.com/dtcenter/MET/develop/met/scripts/Rscripts/pntnc2ascii.R
```

**Reply by PerryShafran-NOAA** [reply]
@hsoh-u I'm trying to find the command Rscript to run the command that you noted above.

**Reply by PerryShafran-NOAA** [reply]
Well I was able to move the file to Hera, where there is an R library, and tried the following:

Rscript /contrib/met/10.1.0-beta6/share/met/Rscripts/pntnc2ascii.R /scratch2/NCEPDEV/fv3-cam/Perry.Shafran/prepbufr.nam.2022030200.nc | grep MLCAPE

And got the following error:

Error in library(ncdf4) : there is no package called 'ncdf4'
Execution halted

Is there anything you can suggest that can help me fix this?

**Reply by hsoh-u** [reply]
Two options but they might not work for you
- System admin installs the missing R packages (because of dependencies)
- Build missing R packages by you and add the path for R package  into ~/.Rprofile

content of ~/.Rprofile:
```
.libPaths("<path_for_your_R_packages>")
```

**Reply by PerryShafran-NOAA** [reply]
I'm pretty sure I don't have the authority to do that.  Is there another module I should be loading perhaps?  I'm not entirely sure why I would be getting this error.  I'm pretty sure that Rscript has been run on Hera by other people.

**Reply by PerryShafran-NOAA** [reply]
@JohnHalleyGotway @jprestop Do you have any idea of what might be happening?  Alternatively, could one of you do the Rscript for me on the file in question?  It would help both of us a lot and save some time on trying to figure this out.

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA, I took a look and here's what I found:

1. On hera, I ran into the same issue with R - no "ncdf4" package. It should be easy to install that locally in your home directory but when I tried to, it just hung:
```
R
install.packages("ncdf4");
# this just hangs
```

2. I did copy the NetCDF file to my machine and ran that command to find that there's no MLCAPE obs in that file:
```
Rscript met/scripts/Rscripts/pntnc2ascii.R prepbufr.nam.2022030200.nc | grep MLCAPE | wc -l
> 0
```

3. You can also run this ncdump command on hera to see the list of obs variables present in that file:
```
ncdump -v obs_var /scratch2/NCEPDEV/fv3-cam/Perry.Shafran/prepbufr.nam.2022030200.nc
 obs_var =
  "QOB",
  "TMP",
  "ZOB",
  "UGRD",
  "VGRD",
  "RH",
  "CAPE",
  "PBL",
  "PRMSL",
  "DPT",
  "PWAT",
  "TCDC",
  "VIS",
  "HGT",
  "GUST" ;
```

4. You could also see this by running plot_point_obs:
```
plot_point_obs /scratch2/NCEPDEV/fv3-cam/Perry.Shafran/prepbufr.nam.2022030200.nc mlcape.ps -obs_var MLCAPE
DEBUG 2: Finished plotting 0 locations for mlcape.ps.
```

So definitely no obs of MLCAPE in that file. Is there a command on WCOSS at which you'd like me to take a closer look?

**Reply by PerryShafran-NOAA** [reply]
You, know, now that I think about it, this is the 2nd run of PB2NC without MLCAPE in it so I could convince myself that the other thing that Howard had talked about was true:

This part:

Well I see that for the CAPE case only I have:

DEBUG 3: Derived CAPE = 95 Zero = 69

And for the CAPE and MLCAPE case, I have:

DEBUG 3: Derived CAPE = 190 Zero = 147

This is the non MLCAPE case.  Previously I *did* do an ncdump on that file with MLCAPE and did see MLCAPE on the list.  So let's try again for the MLCAPE case and look at that file to see that there are MLCAPE obs.

**Reply by PerryShafran-NOAA** [reply]
@JohnHalleyGotway Try copying the file (on WCOSS) /gpfs/dell2/emc/verification/noscrub/Perry.Shafran/metplus_camtest/cam/conus_cam/prepbufr.nam.2022030300.nc and doing the Rscript.  I could also pass that to Hera if that's helpful.

**Reply by PerryShafran-NOAA** [reply]
@JohnHalleyGotway Also, the following:

ncdump -v obs_var prepbufr.nam.2022030300.nc

 obs_var =
  "QOB",
  "TMP",
  "ZOB",
  "UGRD",
  "VGRD",
  "RH",
  "CAPE",
  "MLCAPE",
  "PBL",
  "PRMSL",
  "DPT",
  "PWAT",
  "TCDC",
  "VIS",
  "HGT",
  "GUST" ;

MLCAPE is there.

Also:

plot_point_obs prepbufr.nam.2022030300.nc mlcape.ps -obs_var MLCAPE

DEBUG 2: Finished plotting 96 locations.

Seems to look good!

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA, OK, I logged on to WCOSS, posted that file to our [anonymous FTP site](ftp://ftp.rap.ucar.edu/incoming/irap/met_help/shafran_data/prepbufr.nam.2022030300.nc), downloaded it locally, and ran:
```
Rscript pntnc2ascii.R prepbufr.nam.2022030300.nc | egrep MLCAPE | wc -l
> 96
```

So there are 96 MLCAPE records present. Then I ran this command to stash all instance of CAPE and MLCAPE:
```
Rscript pntnc2ascii.R prepbufr.nam.2022030300.nc | egrep CAPE > prepbufr.nam.2022030300_ALL_CAPE.txt
```

And here's the resulting file:
[prepbufr.nam.2022030300_ALL_CAPE.txt](https://github.com/dtcenter/METplus/files/8188006/prepbufr.nam.2022030300_ALL_CAPE.txt)
 
So now that you're demonstrate that PB2NC can derive MLCAPE, the much more difficult question is, are those derived values actually correct? In case it helps, here's the complete dump of all the obs in that file:
[prepbufr.nam.2022030300.txt.gz](https://github.com/dtcenter/METplus/files/8188015/prepbufr.nam.2022030300.txt.gz)

Please let me know if there's any other support you need in answering that question.

**Reply by PerryShafran-NOAA** [reply]
@JohnHalleyGotway I'd like to know what each of the columns mean.  Seems like the last column is the value of CAPE or MLCAPE.  

In a quick look at the values, they seem to make sense.  I would guess that CAPE and MLCAPE don't have to be that close since they are a measure of different parcels.  On the surface I'd say yes, and it would be good to compare them to the model using PointStat.

And that gets me to my next question regarding the PointStat settings.  The grib2 record I would need to read is:

292:15507077:vt=2022030301:90-0 mb above ground:2 hour fcst:CAPE Convective Available Potential Energy [J/kg]

(Assuming that PB2NC makes its calculation in the 90-0 mb above ground layer.  Howard says that this is the case so I'll go with that.)  This is the MLCAPE, and there is no grib record called MLCAPE - this one is it.  How would I set that up in a PointStat METplus config?

Thanks!!

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA, the output of the pntnc2ascii.R script matches the default ASCII input format for ascii2nc. That's in the user's guide, or you can see it in the ascii2nc usage statement:
```
ascii2nc
	The "met_point" ASCII format consists of 11 columns:
		Message_Type Station_ID Valid_Time(YYYYMMDD_HHMMSS)
		Lat(Deg North) Lon(Deg East) Elevation(msl)
		Var_Name(or GRIB_Code) Level Height(msl or agl)
		QC_String Observation_Value
```

I'm trying to find a sample RAP file on WCOSS I can use to make sure I send you good info about configuring Point-Stat. I found the "nam" data references in your METplus configuration file, but I can't find the "rap" data. Where's that?

**Reply by PerryShafran-NOAA** [reply]
You can find RAP in /gpfs/hps/nco/ops/com/rap/prod.  RAP is run on the Cray, so it's not in the same place as NAM.

**Reply by PerryShafran-NOAA** [reply]
NAM should have the 0-90 mb CAPE output also.

**Reply by JohnHalleyGotway** [reply]
Thanks Perry. Ah yes, I must have been looking at the wrong type of file. I did find it in a NAM Hires CONUS nest file.
```
$WGRIB2 /gpfs/dell1/nco/ops/com/nam/prod/nam.20220302/nam.t00z.conusnest.hiresf12.tm00.grib2 | grep CAPE
670:741082208:d=2022030200:CAPE:surface:12 hour fcst:
789:951832772:d=2022030200:CAPE:180-0 mb above ground:12 hour fcst:
794:962038269:d=2022030200:CAPE:90-0 mb above ground:12 hour fcst:
796:963221251:d=2022030200:CAPE:255-0 mb above ground:12 hour fcst:
```

So we want to match the 90-0 mb level in GRIB record 794. I always use plot_data_plane to test out this sort of thing:
```
plot_data_plane /gpfs/dell1/nco/ops/com/nam/prod/nam.20220302/nam.t00z.conusnest.hiresf12.tm00.grib2 ~/nam_mlcape.ps 'name="CAPE"; level="P90-0";' -v 3
...
DEBUG 3: MetGrib2DataFile::data_plane() -> Found exact match for 'CAPE/P90-0' in GRIB2 record 794 field 1 of GRIB2 file '/gpfs/dell1/nco/ops/com/nam/prod/nam.20220302/nam.t00z.conusnest.hiresf12.tm00.grib2'
...
```

And then displaying it:
```
convert -rotate 90 -background white -flatten ~/nam_mlcape.ps ~/nam_mlcape.png
display ~/nam_mlcape.png
```

<img width="693" alt="Screen Shot 2022-03-04 at 12 58 08 PM" src="https://user-images.githubusercontent.com/21087144/156832951-f980ebb4-bc46-4edf-a60d-0f2322e91a65.png">

So just use a level setting of "P90-0".

**Reply by PerryShafran-NOAA** [reply]
So it should look like this?

FCST_VAR17_NAME = CAPE
FCST_VAR17_LEVELS = P90-0
FCST_VAR17_OPTIONS = cnt_thresh = [ >0 ];
FCST_VAR17_THRESH = >500, >1000, >1500, >2000, >3000, >4000
OBS_VAR17_NAME = MLCAPE
OBS_VAR17_LEVELS = L0-100000
OBS_VAR17_OPTIONS = cnt_thresh = [ >0 ]; cnt_logic = UNION;
OBS_VAR17_THRESH = >500, >1000, >1500, >2000, >3000, >4000

The FCST_VAR_NAME says CAPE and not MLCAPE because it is labelled as CAPE in the grib2 file - is that correct?

Thanks!

**Reply by PerryShafran-NOAA** [reply]
OK, this seemed to work.  I think that we are good but I wanted to make sure with folks like Geoff and Logan that the MLCAPE values are OK before proceeding.  We can talk about this at the Monday METplus user meeting.

Perry

**Reply by georgemccabe** [reply]
Hi @PerryShafran-NOAA, I am just following up on this discussion. Did you confirm with Geoff and Logan that the MLCAPE values are OK?

**Reply by PerryShafran-NOAA** [reply]
Yes, we are good on that end!  I apologize for not reporting sooner, but our team is satisfied with the development of MLCAPE in the new version of METplus.

**Reply by georgemccabe** [reply]
Great to hear, thanks!
